pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/automaton/RegExp.java

```diff
@@ -0,0 +1,771 @@
+      }
+      return e;
+    } else {
+      return parseSimpleExp();
+    }
+  }
+
+  final RegExp parseCharClasses()
+      throws IllegalArgumentException {
+    RegExp e = parseCharClass();
+    while (more() && !peek("]")) {
+      e = makeUnion(e, parseCharClass());
+    }
+    return e;
+  }
+
+  final RegExp parseCharClass()
+      throws IllegalArgumentException {
+    char c = parseCharExp();
+    if (match('-')) {
+      if (peek("]")) {
+        return makeUnion(makeChar(c), makeChar('-'));
+      } else {
+        return makeCharRange(c, parseCharExp());
+      }
+    } else {
+      return makeChar(c);
+    }
+  }
+
+  final RegExp parseSimpleExp()
+      throws IllegalArgumentException {
+    if (match('.')) {
+      return makeAnyChar();
+    } else if (check(EMPTY) && match('#')) {
+      return makeEmpty();
+    } else if (check(ANYSTRING) && match('@')) {
+      return makeAnyString();
+    } else if (match('"')) {
+      int start = _pos;
+      while (more() && !peek("\"")) {
+        next();
+      }
+      if (!match('"')) {
+        throw new IllegalArgumentException("expected '\"' at position " + _pos);
+      }
+      return makeString(_inputString.substring(start, _pos - 1));
+    } else if (match('(')) {
+      if (match(')')) {
+        return makeString("");
+      }
+      RegExp e = parseUnionExp();
+      if (!match(')')) {
+        throw new IllegalArgumentException("expected ')' at position " + _pos);
+      }
+      return e;
+    } else if ((check(AUTOMATON) || check(INTERVAL)) && match('<')) {
+      int start = _pos;
+      while (more() && !peek(">")) {
+        next();
+      }
+      if (!match('>')) {
+        throw new IllegalArgumentException("expected '>' at position " + _pos);
+      }
+      String s = _inputString.substring(start, _pos - 1);
+      int i = s.indexOf('-');
+      if (i == -1) {
+        if (!check(AUTOMATON)) {
+          throw new IllegalArgumentException("interval syntax error at position " + (_pos - 1));
+        }
+        return makeAutomaton(s);
+      } else {
+        if (!check(INTERVAL)) {
+          throw new IllegalArgumentException("illegal identifier at position " + (_pos - 1));
+        }
+        try {
+          if (i == 0 || i == s.length() - 1 || i != s.lastIndexOf('-')) {
+            throw new NumberFormatException();
+          }
+          String smin = s.substring(0, i);
+          String smax = s.substring(i + 1);
+          int imin = Integer.parseInt(smin);
+          int imax = Integer.parseInt(smax);
+          int digits;
+          if (smin.length() == smax.length()) {
+            digits = smin.length();
+          } else {
+            digits = 0;
+          }
+          if (imin > imax) {
+            int t = imin;
+            imin = imax;
+            imax = t;
+          }
+          return makeInterval(imin, imax, digits);
+        } catch (NumberFormatException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)